cellmlmanip/singularity_fixes.py

```diff
@@ -0,0 +1,219 @@
+from math import isclose
+
+import networkx as nx
+from sympy import (
+    Abs,
+    Add,
+    Eq,
+    Float,
+    I,
+    Mul,
+    Piecewise,
+    Pow,
+    Wild,
+    exp,
+    log,
+    solveset,
+)
+
+from cellmlmanip.model import FLOAT_PRECISION, Quantity
+
+
+ONE = Quantity(1.0, 'dimensionless')
+
+
+def _generate_piecewise(vs, ve, sp, ex, V):
+    """Generates a piecewsie for expression based on the singularity point sp and vmin (vs) and vmax (ve) """
+    def f(Vx, e):
+        return e.xreplace({V: Vx})
+
+    if vs is None:  # This shouldn't be apiecewise since we have no Vmin / Vmax
+        return ex
+
+    return Piecewise(*[(f(vs, ex) + ((V - vs) / (ve - vs)) * (f(ve, ex) - f(vs, ex)),
+                        Abs(V - sp) < Abs((ve - vs) / 2)), (ex, True)])
+
+
+def _get_U(expr, V, U_offset, exp_function):
+    '''Finds U in ghk equations these are of one of the the following forms where U is an expression over V:
```

Comment:
And why ``` instead of """ ? I'd prefer if we used """ consistently so that I can comment out blocks using ```

Reply:
Can we get a better name than "get_U" ?

Reply:
Also, these might appear in other equations so really would not mention GHK here explicitly.

Reply:
I'm a bit puzzled why you return U without returning what kind of expression it was in? Does it not make a difference for the solution that you apply what the exact form of the equation is?

Reply:
Needs some info about the input and output as well. Seems to return 3 expressions, rather than "U"

Reply:
documentatio updated